framework/decode/vulkan_address_replacer.cpp

```diff
@@ -0,0 +1,791 @@
+            {
+                GFXRECON_LOG_ERROR("VulkanAddressReplacer: input-handle-buffer creation failed");
+            }
+            memcpy(pipeline_context_bda_.input_handle_buffer.mapped_data, addresses_to_replace.data(), num_bytes);
+
+            replacer_params_t replacer_params = {};
+            replacer_params.hashmap.storage   = pipeline_context_bda_.hashmap_storage.device_address;
+            replacer_params.hashmap.size      = hashmap_bda_.size();
+            replacer_params.hashmap.capacity  = hashmap_bda_.capacity();
+
+            // in-place
+            replacer_params.input_handles  = pipeline_context_bda_.input_handle_buffer.device_address;
+            replacer_params.output_handles = pipeline_context_bda_.input_handle_buffer.device_address;
+
+            replacer_params.num_handles = addresses_to_replace.size();
+
+            device_table_->CmdBindPipeline(command_buffer_info->handle, VK_PIPELINE_BIND_POINT_COMPUTE, pipeline_bda_);
+
+            // NOTE: using push-constants here requires us to re-establish the previous data, if any
+            device_table_->CmdPushConstants(command_buffer_info->handle,
+                                            pipeline_layout_,
+                                            VK_SHADER_STAGE_COMPUTE_BIT,
+                                            0,
+                                            sizeof(replacer_params_t),
+                                            &replacer_params);
+            // run a single workgroup
+            constexpr uint32_t wg_size = 32;
+            device_table_->CmdDispatch(command_buffer_info->handle, div_up(replacer_params.num_handles, wg_size), 1, 1);
+
+            // post memory-barrier
+            for (const auto& buf : buffer_set)
+            {
+                barrier(command_buffer_info->handle,
+                        buf,
+                        VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT,
+                        VK_ACCESS_SHADER_WRITE_BIT,
+                        VK_PIPELINE_STAGE_RAY_TRACING_SHADER_BIT_KHR,
+                        VK_ACCESS_SHADER_READ_BIT);
+            }
+
+            // set previous push-constant data
+            if (!command_buffer_info->push_constant_data.empty())
+            {
+                device_table_->CmdPushConstants(command_buffer_info->handle,
+                                                command_buffer_info->push_constant_pipeline_layout,
+                                                command_buffer_info->push_constant_stage_flags,
+                                                0,
+                                                command_buffer_info->push_constant_data.size(),
+                                                command_buffer_info->push_constant_data.data());
+            }
+
+            // mark end of injected commands
+            decode::EndInjectedCommands();
+        } // !hashmap_bda_.empty()
+    }
+}
+
+void VulkanAddressReplacer::init_pipeline()
```

Comment:
For each of these functions, you might want to put a note that they should only be used in a "mark injected" section.  This just makes sure others going forward continue to be consistent

Reply:
alternatively, instead of putting notes to the private methods,
maybe document this ("all vulkan-commands will be wrapped with markers")
as general contract or "expected behavior" inside a class doc-string?

Reply:
I wrote a class-doc saying: 

`Important note: all internal Vulkan-API calls performed by this class are expected to be wrapped by calls to:
 * - decode::BeginInjectedCommands() / decode::EndInjectedCommands()`